eventauth/eventauth.go

```diff
@@ -362,3 +368,199 @@ func (e *eventAllower) commonChecks(event Event) error {
 
 	return nil
 }
+
+// A membershipAllower has the information needed to authenticate a m.room.member event
+type membershipAllower struct {
+	// The user ID of the user who's membership is changing.
+	targetID string
+	// The user ID of the user who sent the membership event.
+	senderID string
+	// The membership of the user who sent the membership event.
+	senderMember memberContent
+	// The previous membership of the user who's membership is changing.
+	oldMember memberContent
+	// The new membership of the user if this event is accepted.
+	newMember memberContent
+	// The m.room.create content for the room.
+	create createContent
+	// The m.room.power_levels content for the room.
+	powerLevels powerLevelContent
+	// The m.room.join_rules content for the room.
+	joinRule joinRuleContent
+}
+
+// newMembershipAllower loads the information needed to authenticate the m.room.member event
+// from the auth events.
+func newMembershipAllower(authEvents AuthEvents, event Event) (m membershipAllower, err error) {
+	if event.StateKey == nil {
+		err = errorf("m.room.member must be a state event")
+		return
+	}
+	m.targetID = *event.StateKey
+	m.senderID = event.Sender
+	if m.create, err = newCreateContentFromAuthEvents(authEvents); err != nil {
+		return
+	}
+	if m.newMember, err = newMemberContentFromEvent(event); err != nil {
+		return
+	}
+	if m.oldMember, err = newMemberContentFromAuthEvents(authEvents, m.targetID); err != nil {
+		return
+	}
+	if m.senderMember, err = newMemberContentFromAuthEvents(authEvents, m.senderID); err != nil {
+		return
+	}
+	if m.powerLevels, err = newPowerLevelContentFromAuthEvents(authEvents, m.create.Creator); err != nil {
+		return
+	}
+	if m.newMember.Membership == "join" {
+		if m.joinRule, err = newJoinRuleContentFromAuthEvents(authEvents); err != nil {
+			return
+		}
+	}
+	return
+}
+
+// membershipAllowed checks whether the membership event is allowed
+func (m *membershipAllower) membershipAllowed(event Event) error {
+	if m.create.roomID != event.RoomID {
+		return errorf("create event has different roomID: %q != %q", event.RoomID, m.create.roomID)
+	}
+	if err := m.create.userIDAllowed(m.senderID); err != nil {
+		return err
+	}
+	if err := m.create.userIDAllowed(m.targetID); err != nil {
+		return err
+	}
+	// Special case the first join event in the room to allow the creator to join.
+	// https://github.com/matrix-org/synapse/blob/v0.18.5/synapse/api/auth.py#L328
+	if m.targetID == m.create.Creator &&
+		m.newMember.Membership == "join" &&
+		m.senderID == m.targetID &&
+		len(event.PrevEvents) == 1 {
+		// prev_events is a list of 2 element lists of event IDs and hashes.
+		if len(event.PrevEvents[0]) != 2 {
+			return errorf("unparsable prev event")
```

Comment:
Why on earth does the `membershipAllower` need to know how to parse `PrevEvents`? I'd MUCH rather if there is an `event.PrevEvents` that it was validated/parsed when it was assigned to `event`, not randomly here.